### PR TITLE
fix(crossseed): relax discovery matching

### PR DESCRIPTION
**Pull request by s0up4200**
Split cross-seed release matching into strict identity checks and looser discovery checks. RSS automation, webhook prechecks, and remote search candidate filtering now tolerate missing groups and some title noise without widening dedup, local match detection, or indexer filtering.

## Summary by CodeRabbit

* **Refactor**
  * Enhanced release-matching with a discovery-aware pathway that optionally relaxes group/site, title, resolution and metadata checks during candidate selection, search and webhook filtering; adds context-aware title normalization, resolution inference (disc-layout aware) and safeguards for HDR/disc edge cases while preserving strict non-discovery behavior.

* **Tests**
  * Added unit tests covering discovery scenarios: relaxed title/region handling, group compatibility, spinoff rejection, source/version boundaries, resolution/HDR inference and context-aware cases.

### CODE REVIEW

**Comment by Audionut** [comment]
```patch
diff --git a/internal/services/crossseed/matching.go b/internal/services/crossseed/matching.go
index 9fe3c544..c57b2a0c 100644
--- a/internal/services/crossseed/matching.go
+++ b/internal/services/crossseed/matching.go
@@ -116,17 +116,27 @@ func (k releaseKey) String() string {
 	return fmt.Sprintf("%d|%d|%d|%d|%d", k.series, k.episode, k.year, k.month, k.day)
 }
 
+type resolutionMatchContext struct {
+	discLayout bool
+	discMarker string
+	rawName    string
+}
+
 // releasesMatch keeps the historical strict identity semantics used by local
 // match views, deduplication, and other places that must avoid widening.
 func (s *Service) releasesMatch(source, candidate *rls.Release, findIndividualEpisodes bool) bool {
-	return s.releasesMatchWithDiscovery(source, candidate, findIndividualEpisodes, false)
+	return s.releasesMatchWithDiscovery(source, candidate, findIndividualEpisodes, false, resolutionMatchContext{}, resolutionMatchContext{})
 }
 
 func (s *Service) releasesMatchDiscovery(source, candidate *rls.Release, findIndividualEpisodes bool) bool {
-	return s.releasesMatchWithDiscovery(source, candidate, findIndividualEpisodes, true)
+	return s.releasesMatchWithDiscovery(source, candidate, findIndividualEpisodes, true, resolutionMatchContext{}, resolutionMatchContext{})
+}
+
+func (s *Service) releasesMatchDiscoveryWithContext(source, candidate *rls.Release, findIndividualEpisodes bool, sourceCtx, candidateCtx resolutionMatchContext) bool {
+	return s.releasesMatchWithDiscovery(source, candidate, findIndividualEpisodes, true, sourceCtx, candidateCtx)
 }
 
-func (s *Service) releasesMatchWithDiscovery(source, candidate *rls.Release, findIndividualEpisodes bool, allowDiscoveryRelaxations bool) bool {
+func (s *Service) releasesMatchWithDiscovery(source, candidate *rls.Release, findIndividualEpisodes bool, allowDiscoveryRelaxations bool, sourceCtx, candidateCtx resolutionMatchContext) bool {
 	if source == candidate {
 		return true
 	}
@@ -212,7 +222,7 @@ func (s *Service) releasesMatchWithDiscovery(source, candidate *rls.Release, fin
 	}
 
 	if allowDiscoveryRelaxations {
-		if !discoveryMetadataMatch(s, source, candidate) {
+		if !discoveryMetadataMatch(s, source, candidate, sourceCtx, candidateCtx) {
 			return false
 		}
 	} else {
@@ -261,26 +271,8 @@ func (s *Service) releasesMatchWithDiscovery(source, candidate *rls.Release, fin
 			return false
 		}
 
-		// Resolution must match (1080p vs 2160p are different files).
-		// Exception: empty resolution is allowed to match SD resolutions (480p, 576p, SD).
-		sourceRes := normalizer.Normalize(source.Resolution)
-		candidateRes := normalizer.Normalize(candidate.Resolution)
-		if sourceRes != candidateRes {
-			// rls omits resolution for many SD releases (e.g. "WEB" without "480p"), so
-			// treat an empty resolution as a match only when the other side is clearly SD.
-			isKnownSD := func(res string) bool {
-				switch normalizeVariant(res) {
-				case "480P", "576P", "SD":
-					return true
-				default:
-					return false
-				}
-			}
-
-			sdFallbackAllowed := (sourceRes == "" && isKnownSD(candidateRes)) || (candidateRes == "" && isKnownSD(sourceRes))
-			if !sdFallbackAllowed {
-				return false
-			}
+		if !resolutionsCompatible(normalizer, source, candidate, sourceCtx, candidateCtx) {
+			return false
 		}
 	}
 
@@ -443,7 +435,7 @@ func normalizeDiscoveryTitle(title string) string {
 	return strings.Join(filtered, " ")
 }
 
-func discoveryMetadataMatch(s *Service, source, candidate *rls.Release) bool {
+func discoveryMetadataMatch(s *Service, source, candidate *rls.Release, sourceCtx, candidateCtx resolutionMatchContext) bool {
 	normalizer := s.stringNormalizer
 	if normalizer == nil {
 		normalizer = stringutils.DefaultNormalizer
@@ -465,22 +457,8 @@ func discoveryMetadataMatch(s *Service, source, candidate *rls.Release) bool {
 		return false
 	}
 
-	sourceRes := normalizer.Normalize(source.Resolution)
-	candidateRes := normalizer.Normalize(candidate.Resolution)
-	if sourceRes != candidateRes {
-		isKnownSD := func(res string) bool {
-			switch normalizeVariant(res) {
-			case "480P", "576P", "SD":
-				return true
-			default:
-				return false
-			}
-		}
-
-		sdFallbackAllowed := (sourceRes == "" && isKnownSD(candidateRes)) || (candidateRes == "" && isKnownSD(sourceRes))
-		if !sdFallbackAllowed {
-			return false
-		}
+	if !resolutionsCompatible(normalizer, source, candidate, sourceCtx, candidateCtx) {
+		return false
 	}
 
 	sourceVersion := normalizer.Normalize(source.Version)
@@ -512,6 +490,60 @@ func discoveryReleaseGroupCompatible(normalizer *stringutils.Normalizer[string,
 	return strings.HasPrefix(sourceNorm, candidateNorm) || strings.HasPrefix(candidateNorm, sourceNorm)
 }
 
+func resolutionsCompatible(normalizer *stringutils.Normalizer[string, string], source, candidate *rls.Release, sourceCtx, candidateCtx resolutionMatchContext) bool {
+	if normalizer == nil {
+		normalizer = stringutils.DefaultNormalizer
+	}
+
+	sourceRes := normalizeVariant(normalizer.Normalize(source.Resolution))
+	candidateRes := normalizeVariant(normalizer.Normalize(candidate.Resolution))
+	if sourceRes == candidateRes {
+		return true
+	}
+
+	// rls omits resolution for many SD releases (e.g. "WEB" without "480p"), so
+	// treat an empty resolution as a match only when the other side is clearly SD.
+	if (sourceRes == "" && isKnownSDResolution(candidateRes)) || (candidateRes == "" && isKnownSDResolution(sourceRes)) {
+		return true
+	}
+
+	// Disc releases often omit an explicit resolution while equivalent
+	// tracker titles advertise the same disc with the correct resolution. Let those through and rely on
+	// the downstream size/file checks to reject non-equivalent matches.
+	if (sourceRes == "" && inferredResolutionFromContext(source, sourceCtx) == candidateRes) ||
+		(candidateRes == "" && inferredResolutionFromContext(candidate, candidateCtx) == sourceRes) {
+		return true
+	}
+
+	return false
+}
+
+func isKnownSDResolution(resolution string) bool {
+	switch normalizeVariant(resolution) {
+	case "480P", "576P", "SD":
+		return true
+	default:
+		return false
+	}
+}
+
+func inferredResolutionFromContext(release *rls.Release, ctx resolutionMatchContext) string {
+	if release == nil {
+		return ""
+	}
+
+	switch {
+	case ctx.discLayout && strings.EqualFold(ctx.discMarker, "BDMV"):
+		titleHint := normalizeVariant(ctx.rawName)
+		if normalizeSource(release.Source) == "UHD.BLURAY" || strings.Contains(titleHint, "UHD") {
+			return "2160P"
+		}
+		return "1080P"
+	default:
+		return ""
+	}
+}
+
 // joinNormalizedSlice converts a string slice to a normalized uppercase string for comparison.
 // Uppercases and joins elements to ensure consistent comparison regardless of case or order.
 func joinNormalizedSlice(slice []string) string {
diff --git a/internal/services/crossseed/matching_discovery_test.go b/internal/services/crossseed/matching_discovery_test.go
index cbd5ddad..2319e140 100644
--- a/internal/services/crossseed/matching_discovery_test.go
+++ b/internal/services/crossseed/matching_discovery_test.go
@@ -249,3 +249,63 @@ func TestReleasesMatchDiscovery_UsesDefaultNormalizerFallback(t *testing.T) {
 		require.True(t, svc.releasesMatchDiscovery(source, candidate, false))
 	})
 }
+
+func TestReleasesMatchDiscovery_AllowsImplicitUHDToMatchExplicit2160p(t *testing.T) {
+	t.Parallel()
+
+	svc := &Service{
+		releaseCache:     NewReleaseCache(),
+		stringNormalizer: stringutils.NewDefaultNormalizer(),
+	}
+
+	source := svc.releaseCache.Parse("Salt.2010.COMPLETE.UHD.BLURAY-COASTER")
+	candidate := svc.releaseCache.Parse("Salt 2010 Theatrical Cut 2160p UHD Blu-ray HEVC TrueHD 7.1-COASTER")
+
+	require.True(t, svc.releasesMatchDiscoveryWithContext(
+		source,
+		candidate,
+		false,
+		resolutionMatchContext{discLayout: true, discMarker: "BDMV", rawName: "Salt.2010.COMPLETE.UHD.BLURAY-COASTER"},
+		resolutionMatchContext{},
+	))
+}
+
+func TestReleasesMatchDiscovery_DoesNotInfer2160pFromPlainBluray(t *testing.T) {
+	t.Parallel()
+
+	svc := &Service{
+		releaseCache:     NewReleaseCache(),
+		stringNormalizer: stringutils.NewDefaultNormalizer(),
+	}
+
+	source := svc.releaseCache.Parse("Salt.2010.COMPLETE.BLURAY-COASTER")
+	candidate := svc.releaseCache.Parse("Salt 2010 2160p Blu-ray HEVC TrueHD 7.1-COASTER")
+
+	require.False(t, svc.releasesMatchDiscoveryWithContext(
+		source,
+		candidate,
+		false,
+		resolutionMatchContext{discLayout: true, discMarker: "BDMV", rawName: "Salt.2010.COMPLETE.BLURAY-COASTER"},
+		resolutionMatchContext{},
+	))
+}
+
+func TestReleasesMatchDiscovery_Infers1080pFromPlainBlurayDiscLayout(t *testing.T) {
+	t.Parallel()
+
+	svc := &Service{
+		releaseCache:     NewReleaseCache(),
+		stringNormalizer: stringutils.NewDefaultNormalizer(),
+	}
+
+	source := svc.releaseCache.Parse("Salt.2010.COMPLETE.BLURAY-COASTER")
+	candidate := svc.releaseCache.Parse("Salt 2010 1080p Blu-ray AVC TrueHD 5.1-COASTER")
+
+	require.True(t, svc.releasesMatchDiscoveryWithContext(
+		source,
+		candidate,
+		false,
+		resolutionMatchContext{discLayout: true, discMarker: "BDMV", rawName: "Salt.2010.COMPLETE.BLURAY-COASTER"},
+		resolutionMatchContext{},
+	))
+}
diff --git a/internal/services/crossseed/service.go b/internal/services/crossseed/service.go
index d4d5c308..b7412b06 100644
--- a/internal/services/crossseed/service.go
+++ b/internal/services/crossseed/service.go
@@ -7335,6 +7335,11 @@ func (s *Service) searchTorrentMatches(ctx context.Context, instanceID int, hash
 	seen := make(map[string]struct{})
 	sizeFilteredCount := 0
 	releaseFilteredCount := 0
+	sourceResolutionCtx := resolutionMatchContext{
+		discLayout: sourceInfo.DiscLayout,
+		discMarker: sourceInfo.DiscMarker,
+		rawName:    sourceTorrent.Name,
+	}
 
 	for _, res := range searchResults {
 		key := res.GUID
@@ -7349,7 +7354,7 @@ func (s *Service) searchTorrentMatches(ctx context.Context, instanceID int, hash
 		}
 
 		candidateRelease := s.releaseCache.Parse(res.Title)
-		if !s.releasesMatchDiscovery(searchRelease, candidateRelease, opts.FindIndividualEpisodes) {
+		if !s.releasesMatchDiscoveryWithContext(searchRelease, candidateRelease, opts.FindIndividualEpisodes, sourceResolutionCtx, resolutionMatchContext{}) {
 			releaseFilteredCount++
 			continue
 		}
```

**Comment by Audionut** [comment]
HDR marker from sites is also tripping up. Not sure how far you want to go with this, and whether ignoring HDR for BDMV content is suitable, falling back onto other markers/size, or whether dropping the HDR marker is better suited in something like the pooling pr.

**Comment by s0up4200** [comment]
> HDR marker from sites is also tripping up. Not sure how far you want to go with this, and whether ignoring HDR for BDMV content is suitable, falling back onto other markers/size, or whether dropping the HDR marker is better suited in something like the pooling pr.

Feel free to commit stuff here @Audionut - unless its all covered by your other PRs 👍

**Comment by s0up4200** [comment]
> HDR marker from sites is also tripping up. Not sure how far you want to go with this, and whether ignoring HDR for BDMV content is suitable, falling back onto other markers/size, or whether dropping the HDR marker is better suited in something like the pooling pr.

Decide on it in the pooling PR 👍 

I committed your suggestion now.

**Comment by Audionut** [comment]
> Decide on it in the pooling PR

Later stages correctly check disc layouts and stop false matching. So this pr is the correct solution imo, since it correctly passes matching results through this stage, when the sourceHDR is missing (as is often the case with BDMV torrent names), but the candidateHDR has the correct HDR type (as is often the case when content is correctly titled on site).

The only downside is more torrent fetching, but the scope is limited to BDMV when the torrent name lacks HDR, and fits the intent of the PR.
